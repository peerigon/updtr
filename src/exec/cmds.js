function installFn(baseCmd) {
    return ({ registry, modules } = {}) =>
        [baseCmd, stringifyRegistry(registry), stringifyModules(modules)].join(
            ""
        );
}

function stringifyModules(modules) {
    return Array.isArray(modules) === true ?
        " " +
              modules
                  // We need to wrap this in double-quotes because some semver
                  // characters like the caret symbol are reserved characters on Windows.
                  .map(({ name, version }) => `"${ name }@${ version }"`)
                  .join(" ") :
        "";
}

function stringifyRegistry(registry) {
    return registry === undefined ? "" : ` --registry "${ registry }"`;
}

function list({ modules } = {}) {
    return [
        "npm ls --json --depth=0",
        Array.isArray(modules) === true ? " " + modules.join(" ") : "",
    ].join("");
}

export default {
    npm: {
        outdated: () => "npm outdated --json --depth=0",
        installMissing: installFn("npm install"),
        install: installFn("npm install"),
        // remove: ({ name }) => ["npm remove ", name].join(""),
        test: () => "npm test",
        list,
    },
    // yarn does not support custom registries yet.
    // However, these renderers accept them anyway.
    yarn: {
        outdated: () => "yarn outdated --json --flat",
        installMissing: installFn("yarn"),
        install: installFn("yarn add"),
        // remove: ({ name }) => ["yarn remove ", name].join(""),
        test: () => "yarn test",
        // We use npm for listing the dependencies because there is no
        // special benefit by using yarn here
        list,
    },
};
