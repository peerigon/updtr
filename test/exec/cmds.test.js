import cmds from "../../src/exec/cmds";

describe("cmds", () => {
    describe(".npm", () => {
        describe(".outdated()", () => {
            it("should match snapshot", () => {
                expect(cmds.npm.outdated()).toMatchSnapshot();
            });
        });
        describe(".installMissing()", () => {
            it("should match snapshot", () => {
                expect(cmds.npm.installMissing()).toMatchSnapshot();
            });
            describe("custom registry", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.npm.installMissing({
                            registry: "http://example.com",
                        })
                    ).toMatchSnapshot();
                });
            });
        });
        describe(".install()", () => {
            describe("one module", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.npm.install({
                            modules: [
                                {
                                    name: "a",
                                    version: "1.0.0",
                                },
                            ],
                        })
                    ).toMatchSnapshot();
                });
                describe("custom registry", () => {
                    it("should match snapshot", () => {
                        expect(
                            cmds.npm.install({
                                registry: "http://example.com/registry",
                                modules: [
                                    {
                                        name: "a",
                                        version: "1.0.0",
                                    },
                                ],
                            })
                        ).toMatchSnapshot();
                    });
                });
            });
            describe("multiple modules", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.npm.install({
                            modules: [
                                {
                                    name: "a",
                                    version: "1.0.0",
                                },
                                {
                                    name: "b",
                                    version: "1.0.0",
                                },
                            ],
                        })
                    ).toMatchSnapshot();
                });
                describe("custom registry", () => {
                    it("should match snapshot", () => {
                        expect(
                            cmds.npm.install({
                                registry: "http://example.com/registry",
                                modules: [
                                    {
                                        name: "a",
                                        version: "1.0.0",
                                    },
                                    {
                                        name: "b",
                                        version: "1.0.0",
                                    },
                                ],
                            })
                        ).toMatchSnapshot();
                    });
                });
            });
        });
        describe(".test()", () => {
            it("should match snapshot", () => {
                expect(cmds.npm.test()).toMatchSnapshot();
            });
        });
        describe(".list()", () => {
            describe("no arguments", () => {
                it("should match snapshot", () => {
                    expect(cmds.npm.list()).toMatchSnapshot();
                });
            });
            describe("with modules", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.npm.list({ modules: ["a", "b", "c"] })
                    ).toMatchSnapshot();
                });
            });
        });
    });
    describe(".yarn", () => {
        describe(".outdated()", () => {
            it("should match snapshot", () => {
                expect(cmds.yarn.outdated()).toMatchSnapshot();
            });
        });
        describe(".installMissing()", () => {
            it("should match snapshot", () => {
                expect(cmds.yarn.installMissing()).toMatchSnapshot();
            });
            describe("custom registry", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.yarn.installMissing({
                            registry: "http://example.com",
                        })
                    ).toMatchSnapshot();
                });
            });
        });
        describe(".install()", () => {
            describe("one module", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.yarn.install({
                            modules: [
                                {
                                    name: "some-module",
                                    version: "1.0.0",
                                },
                            ],
                        })
                    ).toMatchSnapshot();
                });
                describe("custom registry", () => {
                    it("should match snapshot", () => {
                        expect(
                            cmds.yarn.install({
                                registry: "http://example.com/registry",
                                modules: [
                                    {
                                        name: "some-module",
                                        version: "1.0.0",
                                    },
                                ],
                            })
                        ).toMatchSnapshot();
                    });
                });
            });
            describe("multiple modules", () => {
                it("should match snapshot", () => {
                    expect(
                        cmds.yarn.install({
                            modules: [
                                {
                                    name: "a",
                                    version: "1.0.0",
                                },
                                {
                                    name: "b",
                                    version: "1.0.0",
                                },
                            ],
                        })
                    ).toMatchSnapshot();
                });
                describe("custom registry", () => {
                    it("should match snapshot", () => {
                        expect(
                            cmds.yarn.install({
                                registry: "http://example.com/registry",
                                modules: [
                                    {
                                        name: "a",
                                        version: "1.0.0",
                                    },
                                    {
                                        name: "b",
                                        version: "1.0.0",
                                    },
                                ],
                            })
                        ).toMatchSnapshot();
                    });
                });
            });
        });
        describe(".test()", () => {
            it("should match snapshot", () => {
                expect(cmds.yarn.test()).toMatchSnapshot();
            });
        });
        describe(".list()", () => {
            it("should be the same implementation as .npm.list", () => {
                expect(cmds.yarn.list).toBe(cmds.npm.list);
            });
        });
    });
});
