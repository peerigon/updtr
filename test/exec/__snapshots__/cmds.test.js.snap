// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cmds .npm .install() multiple modules custom registry should match snapshot 1`] = `"npm install --registry \\"http://example.com/registry\\" \\"a@1.0.0\\" \\"b@1.0.0\\""`;

exports[`cmds .npm .install() multiple modules should match snapshot 1`] = `"npm install \\"a@1.0.0\\" \\"b@1.0.0\\""`;

exports[`cmds .npm .install() one module custom registry should match snapshot 1`] = `"npm install --registry \\"http://example.com/registry\\" \\"a@1.0.0\\""`;

exports[`cmds .npm .install() one module should match snapshot 1`] = `"npm install \\"a@1.0.0\\""`;

exports[`cmds .npm .installMissing() custom registry should match snapshot 1`] = `"npm install --registry \\"http://example.com\\""`;

exports[`cmds .npm .installMissing() should match snapshot 1`] = `"npm install"`;

exports[`cmds .npm .list() no arguments should match snapshot 1`] = `"npm ls --json --depth=0"`;

exports[`cmds .npm .list() with modules should match snapshot 1`] = `"npm ls --json --depth=0 a b c"`;

exports[`cmds .npm .outdated() should match snapshot 1`] = `"npm outdated --json --depth=0"`;

exports[`cmds .npm .test() should match snapshot 1`] = `"npm test"`;

exports[`cmds .yarn .install() multiple modules custom registry should match snapshot 1`] = `"yarn add --registry \\"http://example.com/registry\\" \\"a@1.0.0\\" \\"b@1.0.0\\""`;

exports[`cmds .yarn .install() multiple modules should match snapshot 1`] = `"yarn add \\"a@1.0.0\\" \\"b@1.0.0\\""`;

exports[`cmds .yarn .install() one module custom registry should match snapshot 1`] = `"yarn add --registry \\"http://example.com/registry\\" \\"some-module@1.0.0\\""`;

exports[`cmds .yarn .install() one module should match snapshot 1`] = `"yarn add \\"some-module@1.0.0\\""`;

exports[`cmds .yarn .installMissing() custom registry should match snapshot 1`] = `"yarn --registry \\"http://example.com\\""`;

exports[`cmds .yarn .installMissing() should match snapshot 1`] = `"yarn"`;

exports[`cmds .yarn .outdated() should match snapshot 1`] = `"yarn outdated --json --flat"`;

exports[`cmds .yarn .test() should match snapshot 1`] = `"yarn test"`;
